File: hono-node/src/schemas/project-schemas.ts
import { z } from "@hono/zod-openapi";

import { PaginationMetaSchema, PaginationQuerySchema } from "@/core/pagination";

export const ProjectSchema = z.object({
  id: z.string().uuid().openapi({ description: "Project ID" }),
  name: z
    .string()
    .openapi({ description: "Project name", example: "Website Redesign" }),
  description: z
    .string()
    .nullable()
    .openapi({
      description: "Project description",
      example: "Complete overhaul",
    }),
  status: z
    .string()
    .openapi({ description: "Project status", example: "active" }),
  ownerId: z
    .string()
    .uuid()
    .nullable()
    .openapi({ description: "Owner user ID" }),
  createdAt: z
    .string()
    .datetime()
    .openapi({ description: "Creation timestamp" }),
  updatedAt: z
    .string()
    .datetime()
    .openapi({ description: "Last update timestamp" }),
});

export const ProjectCreateSchema = z.object({
  name: z.string().min(1).max(255).openapi({
    description: "Project name",
    example: "Website Redesign",
  }),
  description: z.string().max(1000).optional().openapi({
    description: "Project description",
    example: "Complete overhaul of the company website",
  }),
  status: z
    .enum(["active", "planning", "completed", "archived"])
    .optional()
    .default("active")
    .openapi({
      description: "Project status",
      example: "active",
    }),
});

export const ProjectUpdateSchema = z.object({
  name: z.string().min(1).max(255).optional().openapi({
    description: "Project name",
  }),
  description: z.string().max(1000).nullable().optional().openapi({
    description: "Project description",
  }),
  status: z
    .enum(["active", "planning", "completed", "archived"])
    .optional()
    .openapi({
      description: "Project status",
    }),
});

export const ProjectListQuerySchema = z
  .object({
    search: z.string().optional().openapi({
      description: "Search by name",
    }),
    status: z.string().optional().openapi({
      description: "Filter by status",
    }),
  })
  .merge(PaginationQuerySchema);

export const ProjectListResponseSchema = z.object({
  data: z.array(ProjectSchema),
  meta: PaginationMetaSchema,
});

File: fastapi/src/schemas/project_schemas.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from src.core.pagination import PaginationMeta, PaginationQuery


class ProjectResponse(BaseModel):
    id: UUID = Field(description="Project ID")
    name: str = Field(description="Project name", examples=["Website Redesign"])
    description: str | None = Field(default=None, description="Project description", examples=["Complete overhaul"])
    status: str = Field(description="Project status", examples=["active"])
    owner_id: UUID | None = Field(default=None, description="Owner user ID")
    created_at: datetime = Field(description="Creation timestamp")
    updated_at: datetime = Field(description="Last update timestamp")


class ProjectCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255, description="Project name", examples=["Website Redesign"])
    description: str | None = Field(
        default=None,
        max_length=1000,
        description="Project description",
        examples=["Complete overhaul of the company website"],
    )
    status: str = Field(
        default="active",
        description="Project status",
        examples=["active"],
    )


class ProjectUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255, description="Project name")
    description: str | None = Field(default=None, max_length=1000, description="Project description")
    status: str | None = Field(default=None, description="Project status")


class ProjectListQuery(PaginationQuery):
    search: str | None = Field(default=None, description="Search by name")
    status: str | None = Field(default=None, description="Filter by status")


class ProjectListResponse(BaseModel):
    data: list[ProjectResponse]
    meta: PaginationMeta
